server/document.js

```diff
@@ -182,24 +182,25 @@ export class NextScript extends Component {
     const pagePathname = getPagePathname(pathname)
 
     return <Fragment>
-      {staticMarkup ? null : <script nonce={this.props.nonce} dangerouslySetInnerHTML={{
-        __html: `
-          __NEXT_DATA__ = ${htmlescape(__NEXT_DATA__)}
-          module={}
-          __NEXT_LOADED_PAGES__ = []
-
-          __NEXT_REGISTER_PAGE = function (route, fn) {
-            __NEXT_LOADED_PAGES__.push({ route: route, fn: fn })
-          }${page === '_error' ? `
-
-          __NEXT_REGISTER_PAGE(${htmlescape(pathname)}, function() {
-            var error = new Error('Page does not exist: ${htmlescape(pathname)}')
-            error.statusCode = 404
-
-            return { error: error }
-          })`: ''}
-        `
-      }} />}
+      {staticMarkup ? null : <script
+        id="server-app-state"
+        type="application/json"
+        dangerouslySetInnerHTML={{
+          __html: JSON.stringify({ ...__NEXT_DATA__ }).replace(
+            /<\/script>/g,
+            '%3C/script%3E'
+          ),
+        }}
+      />}
+      <script
+        dangerouslySetInnerHTML={{
+          __html: `module={}`
+        }}
+      />
+      <script
+        src={`${assetPrefix}/_next/static/runtime/bootstrap.js`}
```

Comment:
I'd like to see this discussed upfront, since this adds a network request. I'd rather prefer to have the app-state/bootstrap code integrated into some common bundle (`_app`?).

Reply:
It seems like all the pages are loaded asyncronously so they could be loaded before runtime/main.js which is where next is loaded. I added a new file called runtime/bootstrap.js that should be small and won’t be loaded asyncronously and will always run before everything else. This is a potential blocking process, but if we keep the file small and never change it so we can keep a huge cache time.

Reply:
The app-state is data (nothing that is executed) that will change with each request to SSR, but the bootstrap.js can be cached instead of being sent with each html request so it’s actually going to save bytes over time, and if you are using http/2 you can have more requests without any performance degradation.